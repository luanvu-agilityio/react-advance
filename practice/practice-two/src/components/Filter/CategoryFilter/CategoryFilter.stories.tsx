import type { Meta, StoryObj } from '@storybook/react'
import { useState } from 'react'
import { CategoryFilter } from './CategoryFilter'
import type { CategoryProps } from 'types/Filter'

const meta: Meta<typeof CategoryFilter> = {
  title: 'Components/Filter/CategoryFilter',
  component: CategoryFilter,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'light',
    },
  },
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof CategoryFilter>

const sampleCategories: CategoryProps[] = [
  { name: 'Bakery', count: 12 },
  { name: 'Fruits', count: 24 },
  { name: 'Vegetables', count: 32 },
  { name: 'Dairy', count: 18 },
  { name: 'Meat', count: 16 },
]

/**
 * Default CategoryFilter with multiple categories and "Fruits" as active
 */
export const Default: Story = {
  args: {
    categories: sampleCategories,
    activeCategory: 'Fruits',
    onCategoryClick: (categoryName) => {
      console.log(`Category clicked: ${categoryName}`)
    },
  },
}

/**
 * CategoryFilter with no active category
 */
export const NoActiveCategory: Story = {
  args: {
    categories: sampleCategories,
    activeCategory: '',
    onCategoryClick: (categoryName) => {
      console.log(`Category clicked: ${categoryName}`)
    },
  },
}

/**
 * CategoryFilter with a single item
 */
export const SingleCategory: Story = {
  args: {
    categories: [{ name: 'All Products', count: 120 }],
    activeCategory: 'All Products',
    onCategoryClick: () => {},
  },
}

/**
 * CategoryFilter with many categories to show scrolling behavior
 */
export const ManyCategories: Story = {
  args: {
    categories: Array(15)
      .fill(null)
      .map((_, i) => ({
        name: `Category ${i + 1}`,
        count: Math.floor(Math.random() * 50) + 5,
      })),
    activeCategory: 'Category 5',
    onCategoryClick: () => {},
  },
}

/**
 * Interactive CategoryFilter that demonstrates state changes
 */
export const Interactive: Story = {
  render: function InteractiveCategoryFilter() {
    const [activeCategory, setActiveCategory] = useState('Fruits')

    return (
      <CategoryFilter
        categories={sampleCategories}
        activeCategory={activeCategory}
        onCategoryClick={setActiveCategory}
      />
    )
  },
}

/**
 * CategoryFilter with categories that have long names
 */
export const LongCategoryNames: Story = {
  args: {
    categories: [
      {
        name: 'Extremely Long Category Name That Might Break The Layout',
        count: 12,
      },
      { name: 'Another Very Long Category Name To Test Wrapping', count: 24 },
      { name: 'Short', count: 5 },
    ],
    activeCategory: 'Short',
    onCategoryClick: () => {},
  },
}

/**
 * CategoryFilter with large counts
 */
export const LargeCountNumbers: Story = {
  args: {
    categories: [
      { name: 'Popular Category', count: 1250 },
      { name: 'Very Popular', count: 9999 },
      { name: 'Less Popular', count: 42 },
    ],
    activeCategory: 'Popular Category',
    onCategoryClick: () => {},
  },
}

/**
 * CategoryFilter with empty categories array (should not render)
 */
export const EmptyCategories: Story = {
  args: {
    categories: [],
    activeCategory: '',
    onCategoryClick: () => {},
  },
}
