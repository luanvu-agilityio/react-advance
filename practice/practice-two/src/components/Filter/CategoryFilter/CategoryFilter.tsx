import type { CategoryProps } from 'types/Filter'
import {
  SectionTitle,
  CategorySection,
  CategoryItems,
  CategoryItem,
  CategoryName,
  CategoryCount,
} from '../FilterStyles'

interface CategoryFilterProps {
  categories: CategoryProps[]
  activeCategory: string
  onCategoryClick: (categoryName: string) => void
}

export const CategoryFilter = ({
  categories,
  activeCategory,
  onCategoryClick,
}: CategoryFilterProps) => {
  if (categories.length === 0) return null

  return (
    <CategorySection>
      <SectionTitle>Categories</SectionTitle>
      <CategoryItems>
        {categories.map((category) => {
          const isActive = category.name === activeCategory

          return (
            <CategoryItem
              key={category.name}
              onClick={() => {
                onCategoryClick(category.name)
              }}
              $isActive={isActive}
            >
              <CategoryName $isActive={isActive}>{category.name}</CategoryName>
              <CategoryCount>{category.count}</CategoryCount>
            </CategoryItem>
          )
        })}
      </CategoryItems>
    </CategorySection>
  )
}
