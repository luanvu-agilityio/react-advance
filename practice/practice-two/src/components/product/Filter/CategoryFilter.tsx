import type { CategoryProps } from 'types/Filter'
import {
  SectionTitle,
  CategorySection,
  CategoryItems,
  CategoryItem,
  CategoryName,
  CategoryCount,
} from './FilterStyles'

interface CategoryFilterProps {
  categories: CategoryProps[]
  activeCategory: string
  onCategoryClick: (categoryName: string) => void
}

export const CategoryFilter = ({
  categories,
  activeCategory,
  onCategoryClick,
}: CategoryFilterProps) => {
  if (categories.length === 0) return null

  return (
    <CategorySection>
      <SectionTitle>Categories</SectionTitle>
      <CategoryItems>
        {categories.map((category, index) => (
          <CategoryItem
            key={index}
            onClick={() => onCategoryClick(category.name)}
            $isActive={category.name === activeCategory}
          >
            <CategoryName $isActive={category.name === activeCategory}>
              {category.name}
            </CategoryName>
            <CategoryCount>{category.count}</CategoryCount>
          </CategoryItem>
        ))}
      </CategoryItems>
    </CategorySection>
  )
}
