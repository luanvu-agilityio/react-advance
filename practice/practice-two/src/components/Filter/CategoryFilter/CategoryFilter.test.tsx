import { render, screen, fireEvent } from '@testing-library/react'
import { CategoryFilter } from './CategoryFilter'
import type { CategoryProps } from 'types/Filter'

describe('CategoryFilter', () => {
  const mockCategories: CategoryProps[] = [
    { name: 'Bakery', count: 12 },
    { name: 'Fruits', count: 24 },
    { name: 'Vegetables', count: 32 },
  ]

  const mockOnCategoryClick = jest.fn()

  beforeEach(() => {
    mockOnCategoryClick.mockClear()
  })

  it('renders all category options correctly', () => {
    render(
      <CategoryFilter
        categories={mockCategories}
        activeCategory="Fruits"
        onCategoryClick={mockOnCategoryClick}
      />
    )

    // Check if title is rendered
    expect(screen.getByText('Categories')).toBeInTheDocument()

    // Check if all categories are rendered
    expect(screen.getByText('Bakery')).toBeInTheDocument()
    expect(screen.getByText('Fruits')).toBeInTheDocument()
    expect(screen.getByText('Vegetables')).toBeInTheDocument()

    // Check if counts are rendered
    expect(screen.getByText('12')).toBeInTheDocument()
    expect(screen.getByText('24')).toBeInTheDocument()
    expect(screen.getByText('32')).toBeInTheDocument()
  })

  it('calls onCategoryClick with correct category name when clicked', () => {
    render(
      <CategoryFilter
        categories={mockCategories}
        activeCategory="Fruits"
        onCategoryClick={mockOnCategoryClick}
      />
    )

    // Click on the first category (Bakery)
    const bakeryCategory = screen.getByText('Bakery')
    fireEvent.click(bakeryCategory)

    // Check if onCategoryClick was called with the correct category name
    expect(mockOnCategoryClick).toHaveBeenCalledWith('Bakery')
  })

  it('renders nothing when categories array is empty', () => {
    const { container } = render(
      <CategoryFilter
        categories={[]}
        activeCategory=""
        onCategoryClick={mockOnCategoryClick}
      />
    )
    expect(container.firstChild).toBeNull()
  })

  it('displays the correct count for each category', () => {
    render(
      <CategoryFilter
        categories={mockCategories}
        activeCategory="Fruits"
        onCategoryClick={mockOnCategoryClick}
      />
    )

    // Check if each category has the correct count displayed
    const bakeryItem = screen.getByText('Bakery').closest('button')
    const fruitsItem = screen.getByText('Fruits').closest('button')
    const vegetablesItem = screen.getByText('Vegetables').closest('button')

    expect(bakeryItem).toHaveTextContent('12')
    expect(fruitsItem).toHaveTextContent('24')
    expect(vegetablesItem).toHaveTextContent('32')
  })

  it('handles no active category gracefully', () => {
    render(
      <CategoryFilter
        categories={mockCategories}
        activeCategory=""
        onCategoryClick={mockOnCategoryClick}
      />
    )

    // No category should have the active class/styling
    const categoryItems = screen.getAllByRole('button')

    categoryItems.forEach((item) => {
      expect(item).not.toHaveAttribute('data-active', 'true')
    })
  })

  it('matches snapshot', () => {
    const { asFragment } = render(
      <CategoryFilter
        categories={mockCategories}
        activeCategory="Fruits"
        onCategoryClick={mockOnCategoryClick}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
