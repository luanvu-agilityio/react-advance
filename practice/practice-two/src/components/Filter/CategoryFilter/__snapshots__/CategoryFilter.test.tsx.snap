// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryFilter Rendering States renders different states correctly: empty-categories 1`] = `<div />`;

exports[`CategoryFilter Rendering States renders different states correctly: many-categories 1`] = `
<div>
  <div
    class="sc-dKREkF hFpCMy"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Categories
    </h2>
    <div
      class="sc-fWnslK dhhdnw"
    >
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Category 1
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          10
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Category 2
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          20
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Category 3
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          30
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Category 4
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          40
        </span>
      </div>
      <div
        class="sc-dIMoHT dUnVQj"
      >
        <span
          class="sc-iQQCXo czIniZ"
        >
          Category 5
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          50
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Category 6
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          60
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Category 7
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          70
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Category 8
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          80
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Category 9
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          90
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Category 10
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          100
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`CategoryFilter Rendering States renders different states correctly: no-active-category 1`] = `
<div>
  <div
    class="sc-dKREkF hFpCMy"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Categories
    </h2>
    <div
      class="sc-fWnslK dhhdnw"
    >
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Bakery
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          12
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Fruits
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          24
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Vegetables
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          32
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`CategoryFilter Rendering States renders different states correctly: single-category 1`] = `
<div>
  <div
    class="sc-dKREkF hFpCMy"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Categories
    </h2>
    <div
      class="sc-fWnslK dhhdnw"
    >
      <div
        class="sc-dIMoHT dUnVQj"
      >
        <span
          class="sc-iQQCXo czIniZ"
        >
          Solo Category
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          5
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`CategoryFilter Rendering States renders different states correctly: with-active-category 1`] = `
<div>
  <div
    class="sc-dKREkF hFpCMy"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Categories
    </h2>
    <div
      class="sc-fWnslK dhhdnw"
    >
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Bakery
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          12
        </span>
      </div>
      <div
        class="sc-dIMoHT dUnVQj"
      >
        <span
          class="sc-iQQCXo czIniZ"
        >
          Fruits
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          24
        </span>
      </div>
      <div
        class="sc-dIMoHT dMbzyN"
      >
        <span
          class="sc-iQQCXo bKcNPU"
        >
          Vegetables
        </span>
        <span
          class="sc-gDpztx llcUei"
        >
          32
        </span>
      </div>
    </div>
  </div>
</div>
`;
